### PR TITLE
Import autocomplete

### LINKED ISSUE #746: Imports autocomplete
When specifying imports `import Data.<cursor>` no modules are suggested. Forcing completions using `ctrl+space` gives only local suggestions.

Is this not supported?

### PR DESCRIPTION
**Pull request by lorenzo**
closes #746

### CODE REVIEW

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 1]
```diff
@@ -141,16 +159,15 @@ instance ModuleCache CachedCompletions where
...
-#else
-        asNamespace :: ImportDecl name -> ModuleName
-        asNamespace imp = fromMaybe (iDeclToModName imp) (ideclAs imp)
-#endif
+
         -- Full canonical names of imported modules
         importDeclerations = map unLoc limports
 
+        -- The list of all importable Modules from all packages
+        moduleNames = map showModName (GM.listVisibleModuleNames GHC.unsafeGlobalDynFlags)
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: Is getting global dynamic flags ok here?

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 2]
```diff
@@ -245,33 +262,48 @@ instance ModuleCache CachedCompletions where
...
-        filtCompls = Fuzzy.filterBy label ident compls
+        filtCompls = Fuzzy.filterBy label prefixText compls
           where
-            compls = if T.null qualifier
+           compls = if T.null prefixModule
               then unqualCompls
-              else Map.findWithDefault [] qualifier qualCompls
+              else Map.findWithDefault [] prefixModule qualCompls
+
+        insertWithoutPrefix orig = J.detail ?~ orig
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: I'm not sure why this line requires `FlexibleContexts`

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs (hunk 1 above)
Comment (Bubba) [reply src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: Is it possible to use `getDynFlags` in `PluginUtils.hs` instead? https://github.com/haskell/haskell-ide-engine/blob/60d6bd40c295cbd3a7443fa7735d4f47dce8b0cd/src/Haskell/Ide/Engine/Plugin/HieExtras.hs#L55-L58

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs (hunk 2 above)
Comment (Bubba) [reply src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: Would adding a type signature to `insertWithoutPrefix` alleviate the need for it?

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 3]
```diff
@@ -1,8 +1,8 @@
-{-# LANGUAGE CPP                 #-}
```
Comment (Bubba) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: Big +1 for getting rid of this!

**Comment by lorenzo** [comment]
@Bubba I fixed those 2 things

**Comment by Bubba** [comment]
LGTM, tested it out locally and works great!

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 4]
```diff
@@ -141,16 +161,15 @@ instance ModuleCache CachedCompletions where
         showModName :: ModuleName -> T.Text
         showModName = T.pack . moduleNameString
 
-#if __GLASGOW_HASKELL__ >= 802
         asNamespace :: ImportDecl name -> ModuleName
         asNamespace imp = fromMaybe (iDeclToModName imp) (fmap GHC.unLoc $ ideclAs imp)
```
Comment (alanz) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: Using `fmap` is a neat way of avoiding CPP. Must put it in my mental toolbox.

**Pull request merged**